pyro/ops/gaussian.py

```diff
@@ -251,33 +277,49 @@ class AffineNormal:
     """
     def __init__(self, matrix, loc, scale):
         assert loc.shape == scale.shape
+        assert matrix.shape[:-2] == loc.shape[:-1]
+        assert matrix.size(-1) == loc.size(-1)
         self.matrix = matrix
         self.loc = loc
         self.scale = scale
+        self._gaussian = None
 
-    @property
+    @lazy_property
     def batch_shape(self):
         return self.matrix.shape[:-2]
 
     def condition(self, value):
+        if value.size(-1) == self.loc.size(-1):
+            prec_sqrt = self.matrix / self.scale.unsqueeze(-2)
+            precision = matmul(prec_sqrt, prec_sqrt.transpose(-1, -2))
+            delta = (value - self.loc) / self.scale
+            info_vec = matvecmul(prec_sqrt, delta)
+            log_normalizer = (-0.5 * self.loc.size(-1) * math.log(2 * math.pi)
+                              - 0.5 * delta.pow(2).sum(-1) - self.scale.log().sum(-1))
+            return Gaussian(log_normalizer, info_vec, precision)
+        else:
+            return self.to_gaussian().condition(value)
+
+    def left_condition(self, value):
```

Comment:
I made this method because `event_permute` does not preserve a cheap form of AffineNormal. More precisely, given y = x @ M + mvn, we can't represent `x` as a noisy affine of `y` unless M is invertible.